### PR TITLE
feat: add new apis for clients

### PR DESCRIPTION
**Pull request by oxsteins**
### **PR Type**
Enhancement, Tests, Documentation

___

### **Description**
• Replaced deprecated `GetSuggestedEpoch` API with three new endpoints: `GetLastCommitted`, `GetLastAllCommitted`, and `GetValidatorSetMetadata`
• Added comprehensive validator set metadata storage and retrieval functionality with badger repository integration
• Implemented new gRPC handlers for settlement chain epoch tracking and validator metadata queries
• Added end-to-end tests covering all new API endpoints with validation scenarios
• Updated client examples and documentation to reflect the new API usage patterns
• Enhanced data serialization support with JSON tags for `ExtraData` entity
• Integrated metadata saving into the validator set generation workflow

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["GetSuggestedEpoch API"] -- "removed" --> B["Deprecated"]
  C["New APIs"] --> D["GetLastCommitted"]
  C --> E["GetLastAllCommitted"] 
  C --> F["GetValidatorSetMetadata"]
  G["Repository"] -- "stores" --> H["Validator Metadata"]
  I["EVM Client"] -- "provides" --> J["Settlement Chain Epochs"]
  D --> I
  E --> I
  F --> G
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>api.pb.go</strong><dd><code>Remove suggested epoch API and add validator metadata and last </code><br><code>committed APIs</code></dd></summary>
<hr>

internal/gen/api/v1/api.pb.go

• Removed <code>GetSuggestedEpochRequest</code> and <code>GetSuggestedEpochResponse</code> <br>message types and their associated methods<br> • Added new message types: <br><code>GetValidatorSetMetadataRequest</code>, <code>ExtraData</code>, <br><code>GetValidatorSetMetadataResponse</code>, <code>GetLastCommittedRequest</code>, <br><code>GetLastCommittedResponse</code>, <code>GetLastAllCommittedRequest</code>, <br><code>GetLastAllCommittedResponse</code>, and <code>ChainEpochInfo</code><br> • Updated message type <br>indices throughout the file to accommodate the removal and additions<br> • <br>Added new protobuf field definitions for validator set metadata and <br>last committed epoch functionality

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-4a96021655285a4b388deab2d814058b18632e3e29c38a901f56b83fc0bbabf9">+562/-215</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>api_grpc.pb.go</strong><dd><code>Update gRPC service interface with new API methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/gen/api/v1/api_grpc.pb.go

• Removed <code>GetSuggestedEpoch</code> method from the gRPC service interface and <br>implementation<br> • Added three new gRPC methods: <code>GetLastCommitted</code>, <br><code>GetLastAllCommitted</code>, and <code>GetValidatorSetMetadata</code><br> • Updated method name <br>constants and client/server interfaces to reflect the API changes<br> • <br>Added corresponding handler methods and client method implementations <br>for the new endpoints

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-ea37356c9853638493471d6200c936d44f0ad9a8ab5eb17fd9d793a8be2930e3">+130/-50</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>badger_repository_validator_set.go</strong><dd><code>Add validator set metadata storage functionality to repository</code></dd></summary>
<hr>

internal/client/repository/badger/badger_repository_validator_set.go

• Added <code>SaveValidatorSetMetadata</code> and <code>GetValidatorSetMetadata</code> methods <br>for storing and retrieving validator set metadata<br> • Implemented <br><code>validatorSetMetadataDTO</code> struct and helper functions for <br>serialization/deserialization<br> • Added new key generation function <br><code>keyValidatorSetMetadata</code> for badger database storage<br> • Added JSON <br>marshaling/unmarshaling support for validator set metadata persistence

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-faf4b448b8fc65cc49ee94afc9782e1d86f529e24558ea392f39ac867b046243">+81/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>types.go</strong><dd><code>Update client type definitions for new APIs</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

api/client/v1/types.go

• Removed <code>GetSuggestedEpochRequest</code> and <code>GetSuggestedEpochResponse</code> type <br>aliases<br> • Added new type aliases for the new API requests and <br>responses: <code>GetLastCommittedRequest</code>, <code>GetLastAllCommittedRequest</code>, <br><code>GetValidatorSetMetadataRequest</code>, and their corresponding response types<br> <br>• Added new data type aliases: <code>ChainEpochInfo</code> and <code>ExtraData</code>

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-c7f61c46be8c63e91998201d9fce74fdcfd83986cebf878cf6d66f57998282c6">+8/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>valset_generator_uc.go</strong><dd><code>Integrate metadata saving into validator set generation</code>&nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/valset-generator/valset_generator_uc.go

• Added <code>SaveValidatorSetMetadata</code> method to the repository interface<br> • <br>Integrated metadata saving functionality into the validator set <br>processing workflow<br> • Added call to save validator set metadata with <br>extra data and commitment data during valset generation

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-2bf2060bf29b24fb4e5d3d391ace5487ee56009d3f9aaa5917fd990cf3807f67">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>get_last_committed_v1.go</strong><dd><code>Implement GetLastCommitted API handler</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/api-server/get_last_committed_v1.go

• Implemented <code>GetLastCommitted</code> gRPC handler method<br> • Added validation <br>for settlement chain ID and network configuration lookup<br> • Integrated <br>with EVM client to retrieve last committed header epoch and epoch <br>start time<br> • Added proper error handling and response construction <br>with epoch information

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-4250627cb97c3cfa9ddb72ccd1423e8d0f5951248197082add1a8feb9cc8716d">+56/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>get_validatorset_metadata_v1.go</strong><dd><code>Implement GetValidatorSetMetadata API handler</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/api-server/get_validatorset_metadata_v1.go

• Implemented <code>GetValidatorSetMetadata</code> gRPC handler method<br> • Added <br>support for optional epoch parameter with fallback to current epoch<br> • <br>Integrated with repository to retrieve validator set metadata and <br>construct signature request hash<br> • Added proper error handling and <br>response construction with extra data and commitment data

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-fa5c98da698ad1a6d09d0d59155535e5c070f4cfa3285d38fa7789714285d682">+52/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>get_last_all_committed_v1.go</strong><dd><code>Implement GetLastAllCommitted API handler</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/api-server/get_last_all_committed_v1.go

• Implemented <code>GetLastAllCommitted</code> gRPC handler method<br> • Added <br>functionality to retrieve last committed epochs for all settlement <br>chains from network configuration<br> • Integrated with EVM client to get <br>epoch information and start times for each chain<br> • Added proper error <br>handling and response construction with epoch information mapping

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-b6d7fe6ed9c982d307e9d7c6b5fb4c0a2d752c7e03689b820825d7c73087e065">+41/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>app.go</strong><dd><code>Update API server interfaces for new methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/api-server/app.go

• Added <code>GetValidatorSetMetadata</code> method to the repository interface<br> • <br>Added <code>GetLastCommittedHeaderEpoch</code> method to the EVM client interface<br> • <br>Updated interface definitions to support the new API functionality

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-b2c8c5eee668400ec6d5e50059cd926595235be3b685ecb54e512e7ba7c60473">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>entity.go</strong><dd><code>Add JSON serialization support to ExtraData entity</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

core/entity/entity.go

• Added JSON tags to <code>ExtraData</code> struct fields (<code>Key</code> and <code>Value</code>)<br> • <br>Enhanced serialization support for the <code>ExtraData</code> entity used in <br>validator set metadata

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-c76f58ee92ae2683d15cba8895a561ca38c5ea42e937e4d819ec38a73d2df3d2">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>api.proto</strong><dd><code>Add new API methods for settlement chain epoch tracking</code>&nbsp; &nbsp; </dd></summary>
<hr>

api/proto/v1/api.proto

• Added three new RPC methods: <code>GetLastCommitted</code>, <code>GetLastAllCommitted</code>, <br>and <code>GetValidatorSetMetadata</code><br> • Removed deprecated <code>GetSuggestedEpoch</code> RPC <br>method and related message types<br> • Added new message types <br><code>ChainEpochInfo</code>, <code>ExtraData</code>, and related request/response messages<br> • <br>Updated <code>GetCurrentEpochResponse</code> structure and removed duplicate <br>epoch/start_time fields

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-79a6bada081db311847cbac3c2417c4e5b7341f24ccf95c0a0ee73bbfe629514">+108/-63</a></td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>metadata_test.go</strong><dd><code>Add comprehensive E2E tests for new metadata APIs</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

e2e/tests/metadata_test.go

• Added comprehensive end-to-end tests for the new <br><code>GetValidatorSetMetadata</code> API endpoint<br> • Added tests for <br><code>GetLastAllCommitted</code> API endpoint with validation against network <br>configuration<br> • Implemented test scenarios for committed epochs, <br>signature requests, aggregation proofs, and edge cases<br> • Added <br>validation for individual chain queries and consistency checks between <br>different API endpoints

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-41fedcd3898df05772e744dbe74834ac84cf58db10fd285ff75116ab30e78609">+317/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>app_mock.go</strong><dd><code>Update mocks for new API methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/api-server/mocks/app_mock.go

• Added mock method <code>GetValidatorSetMetadata</code> to the repository mock <br>interface<br> • Added mock method <code>GetLastCommittedHeaderEpoch</code> to the EVM <br>client mock interface<br> • Updated mock recorder methods to support the <br>new API functionality for testing

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-86352c5516df269d1c9a16d9f3532212e68f25da91c30c87a0d0849cf88fb63d">+31/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>main.go</strong><dd><code>Update client example to use new last committed API</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

api/client/examples/main.go

• Replaced <code>GetSuggestedEpoch</code> method call with <code>GetLastAllCommitted</code> <br>method<br> • Updated example code to calculate the minimum last committed <br>epoch across all chains<br> • Modified the example to demonstrate the new <br>API usage pattern for determining suggested epochs

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-5a34b0088ffe7fd76c4570ecf6a186ecc73d3f065442fbd8e89c9ecc827a99fe">+13/-10</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.html</strong><dd><code>Update API documentation with new endpoints and message types</code></dd></summary>
<hr>

api/docs/v1/index.html

• Added new message types <code>ChainEpochInfo</code> and <code>ExtraData</code> to the table of <br>contents<br> • Added new API endpoints for getting last committed epochs <br>(<code>GetLastCommitted</code>, <code>GetLastAllCommitted</code>)<br> • Added new validator set <br>metadata endpoint (<code>GetValidatorSetMetadata</code>)<br> • Removed deprecated <br><code>GetSuggestedEpoch</code> endpoint from service methods table

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-9481edd747dc2e43c356bb8feeca03ed977db036ad71c4eb87013ffb47d5ab5a">+296/-51</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>doc.md</strong><dd><code>Add documentation for new API methods and message types</code>&nbsp; &nbsp; </dd></summary>
<hr>

api/docs/v1/doc.md

• Added documentation for new message types <code>ChainEpochInfo</code> and <br><code>ExtraData</code><br> • Added documentation for new API methods <code>GetLastCommitted</code>, <br><code>GetLastAllCommitted</code>, and <code>GetValidatorSetMetadata</code><br> • Removed <br>documentation for deprecated <code>GetSuggestedEpoch</code> method<br> • Updated <br>service method table with new endpoints

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-5d4e174b596e009ed44cdba23424848516f1bd818c10deede917b7918436da91">+149/-28</a></td>

</tr>
</table></details></td></tr><tr><td><strong>Additional files</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>get_suggested_epoch_v1.go</strong></td>
  <td><a href="https://github.com/symbioticfi/relay/pull/181/files#diff-b92d56cf1465ff9f7498dcbd3fb5cc842cae22cdafec707434cba6926e7917d3">+0/-26</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: core/entity/entity.go [hunk 1]
```diff
@@ -520,8 +520,8 @@ type ValidatorSetHeader struct {
 }
 
 type ExtraData struct {
-	Key   common.Hash
-	Value common.Hash
+	Key   common.Hash `json:"key"`
```
Comment (ilyasymbiotic) [inline core/entity/entity.go:523]: Entity layer should not know anything about serialization for repository layer.
It's better to make dto like it's done for all other entities. Moreover later we're moving to protobuf.

File: internal/usecase/api-server/get_last_all_committed_v1.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// GetLastAllCommitted handles the gRPC GetLastAllCommitted request
+func (h *grpcHandler) GetLastAllCommitted(ctx context.Context, _ *apiv1.GetLastAllCommittedRequest) (*apiv1.GetLastAllCommittedResponse, error) {
+	cfg, err := h.cfg.EvmClient.GetConfig(ctx, uint64(time.Now().Unix()))
+	if err != nil {
+		return nil, errors.Errorf("failed to get config: %w", err)
+	}
+
+	epochInfos := make(map[uint64]*apiv1.ChainEpochInfo)
+	for _, chain := range cfg.Settlements {
+		lastCommittedEpoch, err := h.cfg.EvmClient.GetLastCommittedHeaderEpoch(ctx, chain)
```
Comment (ilyasymbiotic) [inline internal/usecase/api-server/get_last_all_committed_v1.go:22]: Don't we want to call in parallel mode?

**Review by ilyasymbiotic — CHANGES_REQUESTED** [review]
